src/Css/Style.php

```diff
@@ -1992,6 +1992,15 @@ protected function set_prop_color($prop, $color)
         $this->_props_computed[$prop] = (is_array($munged_color) ? $munged_color["hex"] : $munged_color);
     }
 
+    public function set_direction($direction)
+    {
+        $this->_props["direction"] = $direction;
+        $this->_prop_cache["direction"] = null;
+        if ($direction === 'rtl') {
+            $this->text_align = 'right';
+        }
```

Comment:
Will this over-ride user-defined alignment styling?

Reply:
Yes that is another problem I noted as well and forgot about. Do we have a chance to detect a text-align: left is set by the user instead our default value or work around it somehow?
To improve it a bit we could do at least:

```
if ($direction === 'rtl') {
     if ($this->text_align === 'left') {
          $this->text_align = 'right';
     }
}
```

Reply:
I should have stepped through it earlier. :) I thought the defaults are set initially but seems to just are taken when requesting text_align.

The following change works just fine if text_align is set in the same attribute as direction.

```
if ($direction === 'rtl') {
    if (!isset($this->_props['text_align'])) { 
        $this->text_align = 'right';
    }
}
```

But inheritance has another issue:

```
<div style="text-align: left">
	<p style="direction: rtl">
		Hello World !
	</p>
</div>
```

Inheritance generally works correct and sets the value on the child before direction is applied. But the newly parsed styles (Stylesheet::_parse_properties) are added on a new Style class. I do not know there that the text-align is inherited.

Reply:
Checking the `_props` array is exactly how I was thinking that could be done.

As for inheritance, that does present a problem. I think instead of setting the value we should add a getter for the `text-align` property. If not set it checks to see if `direction` is "rtl" and returns right in that case.

Reply:
This sounds like a really good idea. I'll try that approach.

Reply:
I pushed a fix to this.
Caused by the way the __get method handles unset props (it sets the default value) I was forced to do a change with more implications. The default value for `text-align` is now `initial` and the default value `left` is set in the `get_text_align` method.
It works in my tests. But I do not fully understand the computing and property cache system if it could have a problem now anywhere.

Reply:
At a high level, when settings and getting styles from the Style class the following rules is used:

- When set, either by the generic setter or the specific property setter, the various property array (_props, _props_computed, _prop_cache) are cleared.
- The property in the _props array is set to the value passed in (the [specified value](https://www.w3.org/TR/CSS2/cascade.html#specified-value)).
- If the property requires a computation it is done in a property-specific setter. The value property in the _props_computed array is set to this value (the [computed value](https://www.w3.org/TR/CSS2/cascade.html#computed-value)).
- Any final calculations on the computed value are done in property-specific getters (the [used value](https://www.w3.org/TR/CSS2/cascade.html#used-value)).
- Unset properties are set to the default when the getter is used.
- When get through the generic setter the returned value in _prop_cache is set.
- Getters are typically only used at render time, after styles have been parsed and applied to elements.

This is all kind of a work in progress. I think the lack of consistency in how properties are set/get due to partial implementation is to blame. I have an idea of how the getter logic can be tweaked to better accommodate unset properties. Shall I commit the change for review? Then I have a suggestion for how to tweak the text-align getter.

Reply:
Sure, commit directly if you can improve it.

Reply:
Will do. I decided to go full out with the implementation of the computed value. I won't add any new logic but I will be moving some calculations into the setters. I did a quick test with what I have now and I believe the styling is working as expected though I have a bit more testing to do. I'll submit that work as a separate PR so you can review before you update this branch.